### ISSUE TITLE
Failing to differentiate jitted `lax._reduce_window_sum` / `lax.reduce_window`

### ISSUE THREAD

**Issue by romanngg** [google/jax#7815]
In this  example I define 4 identical pooling functions,
https://colab.research.google.com/gist/romanngg/fdd96829faaf0a7e666bccc5fd27ba58/pooling_fail.ipynb

```python
p = r.normal(r.PRNGKey(1), (1, 1))
args = (1, 1), (2, 1), [(0, 0), (1, 0)]


def pool_1(p):
  return lax._reduce_window_sum(p, *args)


def pool_2(p):
  return lax.reduce_window(p, np.zeros((), p.dtype), lax.add, *args)


pool_1_jit = jit(pool_1)
pool_2_jit = jit(pool_2)
```

but as you can see differentiating their jitted versions doesn't work, producing different errors, e.g.
for `lax.reduce_window`:
```python
NotImplementedError: Differentiation rule for 'reduce_window' not implemented
```
or for `lax._reduce_window_sum`:
```python
RuntimeError                              Traceback (most recent call last)
/usr/local/lib/python3.7/dist-packages/jax/interpreters/xla.py in backend_compile(backend, built_c, options)
    383   # we use a separate function call to ensure that XLA compilation appears
    384   # separately in Python profiling results
--> 385   return backend.compile(built_c, compile_options=options)
    386 
    387 def _execute_compiled_primitive(prim, compiled, result_handler, *args):

RuntimeError: Invalid argument: Window dimensions {
  size: 1
  stride: 1
  window_dilation: 1
  base_dilation: 1
}
dimensions {
  size: 1
  stride: 1
  window_dilation: 1
  base_dilation: 1
}
dimensions {
  size: 1
  stride: 1
  padding_low: -1
  window_dilation: 1
  base_dilation: 1
}
 has a negative low padding.
	, for instruction %reduce-window = f32[2,1,1]{2,1,0} reduce-window(f32[2,1,2]{2,1,0} %parameter.1, f32[] %constant.5), window={size=1x1x1 pad=0_0x0_0x-1_0}, to_apply=%primitive_computation_add.6, metadata={op_type="reduce_window_sum" op_name="jit(vmap(transpose(jvp(pool_1))))/reduce_window_sum[ base_dilation=(1, 1, 1)\n                                                     padding=((0, 0), (0, 0), (0, 0))\n                                                     window_dilation=(1, 1, 1)\n                                                     window_dimensions=(1, 1, 1)\n                                                     window_strides=(1, 1, 1) ]" source_file="<ipython-input-2-ee69b86aa221>" source_line=6}

Failed after simplification
```

**Comment by froystig** [comment]
The first error has all of the information:
```
NotImplementedError: Differentiation rule for 'reduce_window' not implemented
```
We need to implement AD for this primitive.

Regarding the second error, we should always try to avoid generating invalid HLO that results in XLA runtime errors, but note that this one came about by differentiating a non-public function from jax internals, which isn't allowed.

One idea there is to validate window dimension in the jaxpr checker, so that we catch this before lowering to HLO.

**Comment by romanngg** [comment]
Thanks Roy, maybe naive question, but shouldn't this rule be already implemented if I'm able to differentiate it without jit (and it only throws the NotImplementedError if I jit it), so I wonder if there's some bug in interacting between jit and AD?

**Comment by froystig** [comment]
Thanks for highlighting that! I overlooked it in the notebook that you shared.

What's happening is a bit tricky, but has to do with how we handle a special case of `lax.reduce_window` as an optimization just before binding the underlying primitive (which has no AD rule):

https://github.com/google/jax/blob/e7e5140dc9fd192154e866794424d073e5c8d2b0/jax/_src/lax/lax.py#L1387-L1393

This tries to identify known monoids (an associative operation with an identity element) as the reduction function and initial value. For instance, if we successfully notice a reduction with `(+, 0)`, then we proceed to call an internal `_reduce_window_sum`, rather than following the general path. The sum-specific path has AD rules.

When jax transforms a function, it interprets it with abstract arguments. The degree of abstraction depends on the transformation involved. For autodiff alone, we still have enough information to concretely identify the `0` part of the `(+, 0)`, but once we involve compilation as well, we abstract the computation further (to shapes/dtypes) and can at most say that we have addition with a float32 scalar initial value, i.e. "`(+, f32[])`." Since we can't guarantee the `0` initial value, we continue on the general path, for which there is no AD rule.

Note also that, if you were to reduce with an operation other than `+`, this would also send you down the general path, and hence would err with or without `jit`.

I was about to suggest, as a workaround, to hard-code the initial value as a constant zero by rewriting:

```python
import numpy as onp

def pool_2(p):
  return lax.reduce_window(p, onp.zeros((), p.dtype), lax.add, *args)

pool_2_jit = jit(pool_2)
print(jacobian(pool_2_jit)(p))
```

But it turns out this actually reproduces your original XLA runtime error, from differentiating `pool_1_jit`, using only public `lax` functions! We must be hitting the same bug here that you encountered when differentiating `_reduce_window_sum` directly. In hindsight, your code set up that test case correctly, even if calling internal functions isn't otherwise allowed. Apologies there!

In summary, we have a feature request (for AD on the general path) and a bug (with staging+AD on special path) at the same time.

**Comment by froystig** [comment]
I posted #7860 for the bug and we can keep this as the feature request.

**Comment by romanngg** [comment]
Thanks a lot for such detailed explanation Roy!

**Comment by deasmhumhna** [comment]
Has there been any progress on this? I wrote this quick implementation of a `jvp` rule utilizing the fact that `reduce_window` can take multiple `operands`. I first tried setting it as the rule for the `reduce_window_p` primitive but couldn't figure out how to get the function argument passed in `params` (#4918). Still need to implement a transpose rule, though an explicit `vjp` may be more straightforward as a `scatter_add` of the `computation`-level `vjp`  into the original shape. 
```python3
from typing import Any, Callable, Sequence, Union, Optional

import jax 
import jax.lax as lax 
import jax.numpy as jnp

from functools import partial

Array = Any

@partial(jax.custom_jvp, nondiff_argnums=range(2, 8))
def reduce_window(
    operand: Array, 
    init_value: Array,
    computation: Callable,
    window_dimensions: Sequence[Union[int, Any]], 
    window_strides: Sequence[int], 
    padding: Union[str, Sequence[tuple[int, int]]],
    base_dilation: Optional[Sequence[int]]=None, 
    window_dilation: Optional[Sequence[int]]=None
):
    return lax.reduce_window(
        operand=operand,
        init_value=init_value,
        computation=computation, 
        window_dimensions=window_dimensions, 
        window_strides=window_strides, 
        padding=padding,
        base_dilation=base_dilation, 
        window_dilation=window_dilation
    )

@reduce_window.defjvp
def reduce_window_jvp(
    computation: Callable,
    window_dimensions: Sequence[Union[int, Any]], 
    window_strides: Sequence[int], 
    padding: Union[str, Sequence[tuple[int, int]]],
    base_dilation: Optional[Sequence[int]], 
    window_dilation: Optional[Sequence[int]],
    primals: Array, 
    tangents: Array
):
    def fun(*args):
        primals, tangents = zip(*args)
        return jax.jvp(computation, primals, tangents)

    def make_zero(tan):
        return lax.zeros_like_array(x) if type(tan) is ad.Zero else tan 
    
    tangents = tuple(make_zero(tan) for tan in tangents)

    primals, tangents = lax.reduce_window(
        operand=(primals[0], tangents[0]),
        init_value=(primals[1], tangents[1]),
        computation=fun, 
        window_dimensions=window_dimensions, 
        window_strides=window_strides, 
        padding=padding,
        base_dilation=base_dilation, 
        window_dilation=window_dilation
    )
    return primals, tangents

x = jnp.linspace(-12., 12., 25).reshape(5, 5)
x_dot = jnp.ones_like(x)
test = reduce_window_jvp(lax.add, (3, 3), (1, 1), ((1, 1), (1, 1)), None, None, (x, 0.), (x_dot, 0.))
true = jax.jvp(lambda x: lax.reduce_window(x, 0., lax.add, (3, 3), (1, 1), ((1, 1), (1, 1))), (x,), (x_dot,))
print([jnp.allclose(elem1, elem2) for (elem1, elem2) in zip(test, true)]) #[Array(True, dtype=bool), Array(True, dtype=bool)]
```

**Comment by deasmhumhna** [comment]
I realized the `computation` function is passed as a Jaxpr and created a jvp-rule for the `reduce_window` primitive. I've also implemented `custom_vjp` that I should be able to adapt into a transposition operation for general pytrees. However, it seems that `linearize` requires that the primal values be partially evaluated without knowing the tangents, which is not possible for this op. Is this a computation requirement or a convention? 

```python3
def accumulator(computation):
    def fun(x, y):
        z = computation(x[0], y[0])
        index = jnp.array(x[1], dtype=jnp.int32)
        rest = jnp.stack(x[2:], axis=0)
        rest = jnp.where(jnp.arange(len(rest)) == index, z, rest)
        rest = jnp.where(jnp.arange(len(rest)) == len(rest) // 2 + index, y[1], rest)
        rest = jnp.split(rest, len(rest))
        rest = [elem.squeeze() for elem in rest]
        return z, index + 1., *rest
    return fun


def reduce_window_fwd(
    operand: Array, 
    init_value: Array,
    computation: Callable,
    window_dimensions: Sequence[Union[int, Any]], 
    window_strides: Sequence[int], 
    padding: Union[str, Sequence[tuple[int, int]]],
    base_dilation: Optional[Sequence[int]]=None, 
    window_dilation: Optional[Sequence[int]]=None
):
    size = reduce(mul, window_dimensions)
    location = jnp.arange(operand.size, dtype=operand.dtype).reshape(operand.shape)
    new_operand = (operand, location) + (jnp.zeros_like(operand),) * (2 * size)
    new_init_value = (init_value, 0.) + (0.,) * size + (-1.,) * size

    out, _, *intermediates = lax.reduce_window(
        operand=new_operand,
        init_value=new_init_value,
        computation=accumulator(computation), 
        window_dimensions=window_dimensions, 
        window_strides=window_strides, 
        padding=padding,
        base_dilation=base_dilation, 
        window_dilation=window_dilation
    )
    return out, (operand, init_value, intermediates)


def reduce_window_bwd(
    computation: Callable,
    window_dimensions: Sequence[Union[int, Any]], 
    window_strides: Sequence[int], 
    padding: Union[str, Sequence[tuple[int, int]]],
    base_dilation: Optional[Sequence[int]], 
    window_dilation: Optional[Sequence[int]],
    context, 
    tangents
):
     def step(carry, x):
        tout = carry 
        input1, input2 = x
        _, vjpfun = jax.vjp(computation, input1, input2)
        tin1, tin2 = vjpfun(tout)
        carry, targ = map(
            partial(jnp.where, jnp.isfinite(input2)), [tin1, tin2], [tout, 0.])
        return carry, targ
    
    operand, init_value, intermediates = context
    length = len(intermediates)
    index = jnp.stack(intermediates[length // 2:], axis=0).astype(jnp.int32)
    indices = jnp.unravel_index(index, operand.shape)
    init_value_broadcast = jnp.full(operand.shape, init_value)
    input1 = jnp.stack(
        [init_value_broadcast] + intermediates[:length // 2 - 1], axis=0)
    input2 = jnp.where(index >= 0, operand[indices], jnp.nan)
    t_init, t_operand_stacked = lax.scan(
        step, tangents, xs=(input1, input2), reverse=True)
    t_operand = lax.scatter_add(
        jnp.zeros_like(operand), 
        jnp.stack(indices, axis=-1), 
        updates=t_operand_stacked, 
        dimension_numbers=lax.ScatterDimensionNumbers((), (0, 1), (0, 1))
    )
    return t_operand, t_init.sum()

reduce_window.defvjp(reduce_window_fwd, reduce_window_bwd)

x = jnp.linspace(-12., 12., 25).reshape(5, 5)
test = jax.grad(lambda x, y: jnp.cos(reduce_window(x, y, lax.max, (3, 3), (1, 1), ((1, 1), (1, 1)))).sum())(x, -1e7)
true = jax.grad(lambda x: jnp.cos(lax.reduce_window(x, -jnp.inf, lax.max, (3, 3), (1, 1), ((1, 1), (1, 1)))).sum())(x)
jnp.allclose(test, true) # Array(True, dtype=bool)
```

**Comment by voznesenskym** [comment]
I did the jvp for this. Working on transpose.

**Comment by deasmhumhna** [comment]
> I did the jvp for this. Working on transpose.

Very cool. I think I final did get a jvp and transpose working based on an the inverse op from `oryx`. Honestly forgot why I needed this in the first place but here's the colab: https://colab.research.google.com/drive/1B0o4P1T4UTLI4Q8eKOdzUjI4DnfzR6nP?usp=sharing

**Comment by voznesenskym** [comment]
Oh cool! Yeah the jvp is landed in main as of this week.

Ty for sharing the colab. I’m afraid to click that link. I already have
transpose mostly working (higher orders, and composing nicely with jvp, all
passing grad accuracy checks etc) except without dilation yet (gonna add
that today). If I open your link, I’m gonna just see the answer and lose
out on the fun lol.
